src/spaces/compartment.jl

```diff
@@ -0,0 +1,379 @@
+function move_agent!(agent::A, model::ABM{A, <: CompartmentSpace}, dt::Real = 1.0) where {A <: AbstractAgent}
+    model.space.update_vel!(agent, model)
+    pos = agent.pos .+ dt .* agent.vel
+    move_agent!(agent, pos, model)
+    return agent.pos
+end
+
+#######################################################################################
+# %% Neighbors and stuff
+#######################################################################################
+
+grid_space_neighborhood(α, model::ABM{<:AbstractAgent, <:CompartmentSpace}, r) =
+grid_space_neighborhood(α, model.space.grid, r)
+
+function nearby_ids_cell(pos::ValidPos, model::ABM{<:AbstractAgent,<:CompartmentSpace}, r = 1)
+    nn = grid_space_neighborhood(CartesianIndex(pos2cell(pos)), model, r)
+    s = model.space.grid.s
+    Iterators.flatten((s[i...] for i in nn))
+end
+
+function nearby_positions(pos::ValidPos, model::ABM{<:AbstractAgent,<:CompartmentSpace}, r = 1)
+    nn = grid_space_neighborhood(CartesianIndex(pos2cell(pos)), model, r)
+    Iterators.filter(!isequal(pos), nn)
+end
+
+function positions(model::ABM{<:AbstractAgent,<:CompartmentSpace})
+    x = CartesianIndices(model.space.grid.s)
+    return (Tuple(y) for y in x)
+end
+
+function ids_in_position(pos::ValidPos, model::ABM{<:AbstractAgent,<:CompartmentSpace})
+    return model.space.grid.s[pos...]
+end
+
+cell_center(pos::Tuple) = getindex.(modf.(pos), 2) .+ 0.5
+distance_from_cell_center(pos::Tuple, center) = sqrt(sum(abs2.(pos .- center)))
+
+"""
+        nearby_ids(pos::ValidPos, model::ABM{<:AbstractAgent,<:CompartmentSpace}, r=1; exact=false) → ids
+
+Return an array of the ids of the agents within "radius" `r` of the given `position` in `CompartmentSpace`.
+
+If an agent is given instead of a position `pos`, the id of the agent is excluded.
+
+# Keywords
+* `exact=false` checks for exact distance rather than returing the ids of all
+agents in cells within `r`.
+"""
+function nearby_ids(pos::ValidPos, model::ABM{<:AbstractAgent,<:CompartmentSpace}, r=1; exact=false)
+    center = cell_center(pos)
+    focal_cell = ceil.(Int, center)
+    if exact
+        newr = ceil.(Int, r)
+        corner_of_largest_square_in_circle = floor.(Int, pos .+ newr)
+        max_distance = corner_of_largest_square_in_circle[1] - focal_cell[1]
+        if max_distance == 0 
+            max_distance = r
+        end
+        final_ids = Int[]
+        allcells = grid_space_neighborhood(CartesianIndex(focal_cell), model, newr)
+        for cell in allcells
+            if !any(x-> x> max_distance, abs.(cell .- focal_cell)) && max_distance > 1 # certain cell
+                ids = ids_in_position(cell, model)
+                final_ids = vcat(final_ids, ids)
+            else # uncertain cell
+                ids = copy(ids_in_position(cell, model))
+                filter!(i -> sqrt(sum(abs2.(pos .- model[i].pos))) ≤ r, ids)
+                final_ids = vcat(final_ids, ids)
+            end
+        end
+        return final_ids
```

Comment:
This part of the code has bad performance. Instead of continuously re-creating `final_ids` by vcat-ing it with itself and something else we should follow the [Julia performance tips](https://docs.julialang.org/en/v1/manual/performance-tips/) and reduce allocations. 

The `ids` you find in this loop should be added to `final_ids` using `append!` .